### PR TITLE
`sets` assign operators

### PR DESCRIPTION
**Pull request by narimiran**
Adds assign operators, for `a += b` usage, instead of `a = a + b`.

### CODE REVIEW

**Comment by RedBeard0531** [comment]
@dom96 Wouldn't it be better to go the other way? Usually `a += b` can be implemented more efficiently than `a = a + b`, so it probably makes more sense for `+=` to be the customization point and have a generic `+` that is implemented as copy then `+=`. I know that this patch doesn't take advantage of this, but `A += B` should be implementable in-place with no allocations.

**Comment by narimiran** [comment]
> we should have a single generic += that calls +.

This sounds good to me!

The things @RedBeard0531 mentioned also make sense to me. Your thoughts, @dom96?

**Comment by dom96** [comment]
What @RedBeard0531 says makes sense, however ``+=`` isn't idiomatic Nim. The idiomatic way to increment a int for example is via ``x.inc(23)``. So IMO the following should be done:

* ``+=`` in system.nim should be changed to a template which calls ``x.inc(number)``.
* ``inc`` should be implemented for sets
* The same for the other arithmetic (note this does not include the weird ``-+-``, I don't think this deserves a ``-+-=``)

File: lib/pure/collections/sets.nim [hunk 1]
```diff
@@ -516,6 +516,22 @@ proc `-+-`*[A](s1, s2: HashSet[A]): HashSet[A] {.inline.} =
...
+
+proc `*=`*[A](s1: var HashSet, s2: HashSet[A]) {.inline.} =
+  ## Alias for `s1 = intersection(s1, s2) <#intersection>`_.
+  s1 = intersection(s1, s2)
+
+proc `-=`*[A](s1: var HashSet, s2: HashSet[A]) {.inline.} =
+  ## Alias for `s1 = difference(s1, s2) <#difference>`_.
+  s1 = difference(s1, s2)
+
+proc `-+-=`*[A](s1: var HashSet, s2: HashSet[A]) {.inline.} =
```
Comment (Araq) [inline lib/pure/collections/sets.nim]: I've never used "symmetric set difference" in my entire life. Can we not introduce pointless shortcuts please? Consistency is not the only criterion for a design. Another one is "bloat".

**Comment by RedBeard0531** [comment]
> @dom96 
> however `+=` isn't idiomatic Nim

Is that really the case for non-numeric arguments? I'm still new I can't claim to have a great sense of the idioms, but `inc` on sets just feels weird to me in a way that `+=` doesn't. Would it be reasonable to restrict the `inc` idiom to numeric types?

**Comment by narimiran** [comment]
> `inc` on sets just feels weird to me

...because `+` and `+=` on sets aren't `inc`, but `union` - which in some other languages is represented with `|` and `|=`.

**Comment by Araq** [comment]
> ...because + and += on sets aren't inc, but union - which in some other languages is represented with | and |=.

``+`` for union is as "logical" as ``|`` if not moreso, there is also ``-`` (set *difference*, so there).

File: lib/pure/collections/sets.nim (hunk 1 above)
Comment (narimiran) [reply lib/pure/collections/sets.nim]: After just 3 months, I addressed this :)

**Pull request closed without merging**